### ISSUE TITLE
Make BreakIterator more its icu4c counterpart

### ISSUE THREAD

**Issue by conniey** [sillsdev/icu-dotnet#24]
It would be easier to consume BreakIterator if it was set-up similar to the way it is in icu4c and icu4j, as a class, since it is being passed as method and constructor parameters and because it is meant to be an extension point where you can design your own word breaking if you need to customize the default ICU behavior. 

I want to modify BreakIterator to:

* Be a class rather than static
* Add methods similar to the ones from [icu::BreakIterator](http://icu-project.org/apiref/icu4c/classicu_1_1BreakIterator.html)
* Move the enumerations such as `UBreakIteratorType`, `ULineBreakTag` outside of BreakIterator.

So the public API would look like:

```csharp
using System;

public BreakIterator : IDisposable
{
    // New ctor, methods and properties
    protected BreakIterator(UBreakIteratorType type, Locale locale, string text);

    // includeSpacesAndPunctuation only applies to UBreakIteratorType.Word, so most of the time it is true.
    protected BreakIterator(UBreakIteratorType type, Locale locale, string text, bool includeSpacesAndPunctuation);

    public virtual Boundary Next { get; }
    public virtual Boundary Current { get; }
    public virtual Boundary Previous { get; }
    public virtual Boundary First { get; }
    public virtual Boundary Last { get; }

    public virtual Boundary[] Boundaries { get; }
    public virtual string Text { get; }
    public Locale Locale { get; }

    public virtual void SetText(string text);

    public static BreakIterator CreateCharacterInstance(Locale locale, string text);
    public static BreakIterator CreateWordInstance(Locale locale, string text);
    public static BreakIterator CreateLineInstance(Locale locale, string text);
    public static BreakIterator CreateSentenceInstance(Locale locale, string text);
    public static BreakIterator CreateWordInstance(Locale locale, string text);

    // Existing methods
    public static IEnumerable<string> Split(UBreakIteratorType type, Locale locale, string text);
    public static IEnumerable<string> Split(UBreakIteratorType type, string locale, string text);
    public static IEnumerable<Boundary> GetWordBoundaries(Locale locale, string text, bool includeSpacesAndPunctuation);
    public static IEnumerable<Boundary> GetWordBoundaries(string locale, string text, bool includeSpacesAndPunctuation);
    public static IEnumerable<Boundary> GetBoundaries(UBreakIteratorType type, Locale locale, string text);
}
```

Would you be opposed to this?

**Comment by JohnThomson** [comment]
I don't understand why you want an abstract class. The usual reason for
making a class abstract is so that at least one method can be abstract,
thus forcing clients to subclass and implement that method. But no method
in your proposed API is abstract.

I could see making various methods virtual so it would be possible (but not
required) to subclass it for different behavior.

I could see making it an interface (so using the current implementation
would be optional and the behavior could be entirely replaced).

But I'm not seeing the reason for making it abstract, and a breaking change
should have a good justification.

John Thomson

**Comment by conniey** [comment]
In my original post, I had suggested making it abstract, but I updated the issue several times afterwards and now it is a concrete class now with several virtual methods.

Connie

**Comment by NightOwl888** [comment]
@JohnThomson 

[BreakIterator](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/text/BreakIterator.java) is an abstract class in the JDK. I think it was made abstract because (judging by the comments) it was slipped in after there were already several BreakIterator types that had similar interfaces, and there were some methods that were missing from some of them that needed a default implementation.

It makes sense in this case to make an abstraction, since it doesn't really make much sense to make RuleBasedBreakIterator subclass a concrete BreakIterator type. It is unlikely that much logic will be shared, and worse, the author of the subclass might accidentally overlook one of the virtual method implementations and have a difficult to diagnose bug.

But we could just as well make this abstraction an interface. The only argument that might make a case for an abstract class is because it has a shared constant (DONE).

**Comment by JohnThomson** [comment]
The proposal to make it abstract has been withdrawn. I gather the expectation is that the concrete class (or a subclass) will have non-trivial useful behavior (possibly can be used to implement CreateCharacterInstance etc?) even when nothing is overridden. Overriding is just an optional use case enabled by making appropriate methods virtual.
The important thing, I think, is that because the existing methods were all static, this change shouldn't break any clients.

**Comment by ermshiperete** [comment]
Fixed by PR #25
